### DISCUSSION TITLE
Draw a graph showing charging and discharging of a battery

### DISCUSSION THREAD

**Discussion by rana253** [CURENT/andes#272] [Q&A]
Hi Dr. Cui,
I am working on  **Ieee14** bus system `ESD1 `model.
I have done simulation of model. Now our requirement is to show the charging and discharging of a battery graphically. For this we need to discharge a fully charged battery to a minimum value (P=0) and again charging it to max value. In order to achieve this process kindly advise which parameters or values are to be dynamically changed to get battery charging and discharging.

Regards,
Rana

**Comment by cuihantao** [comment]
Hi,

The variable `pIG_y` of `ESD1` is the state of charge. As the simulation goes, the state of charge will change, depending on the initial power output and the disturbances. Even for a system without disturbance, the SOC should change at a constant rate.

If you assign a capacity `En` that is small enough comparative to the power outputs (obtained from power flow), the battery should discharge to `SOCmin` quickly. 

To convert the battery back to a charging state, the incremental power adjustment `Pext0` can be set. The value of `Pext0` will be added to the total power outputs `Psum`, which will be used to calculate the current injection. See https://github.com/cuihantao/andes/blob/7da7d4d1ac814a3f41266c85bea14fa526b38770/andes/models/distributed/pvd1.py#L369-L391

**Reply by rana253** [reply]
Hi Dr. Cui,
Thank you for your reply.
I have followed the steps as suggested by you
1. I have assigned a capacity En (very small value) and run the system to discharge the battery to the level of SoCmin.
**Now in order to charge the battery again, I changed the value of Pext0 (increased) but the graph is still moving on the same level i.e. SoCmin instead of moving upwards (towards SoCmax)**
2. After 1st attempt, I tried with Constant change in SoC method (without any disturbances)- I made the battery discharge by 10% only and tried to charge it again from that point by incrementing the value of Pext0, but the battery continues to discharge instead of charging.
**My query is-** while charging the battery from SoCmin how the value of Pext0 should be changed so that the battery gets charged again. Is there any reference value of Pext0 to get the desired result?

**Reply by cuihantao** [reply]
Did you post the other discussion here? https://github.com/cuihantao/andes/discussions/273

Like that one, please check `Psum` and see if it gets changed correctly. You need to post your code if you need further diagnostics.

**Comment by rana253** [comment]
Hi Dr. Cui,

I am working on preparation of customized 9 bus model. In this we need to calculate p0 and q0 value for PQ, PV and Slack Bus.
I have done the calculation as described below:
Base MVA = 100
Load= 50kW
50 kW= 0.05 MVA (unity power factor)
PU value of p0 = 0.05 / 100 = 0.0005 
Similarly, calculation done for all the buses
Please suggest if this is the correct method, and if  its not kindly advise how to calculate p0 and q0 for all the buses

**Reply by cuihantao** [reply]
That is correct!

**Reply by rana253** [reply]
Hi Dr. Cui,
Thank you for your response.
I am facing issue while doing simulation, let me describe my system first.
in these system I have considered 
bus 1 slack Genrou 100 MVA
bus 2 PVD1              2.2 MVA
bus 3 ESD1               2.5 MVA
and 6 load in PQ bus

while simulation I am getting below mentioned error

GENROU (vf range) out of typical lower limit.

   idx    | values | limit
----------+--------+------
 GENROU_1 | -0.201 | 1    


PVD1.Iqcmd_lim out of limits <Iqmax>

  idx   | Flag | Input Value | Limit
--------+------+-------------+------
 PVD1_1 | zu   | 0.341       | 0.021


ESD1.Iqcmd_lim out of limits <Iqmax>

idx | Flag | Input Value | Limit
----+------+-------------+------
 1  | zu   | 0.342       | 0.027


Suspect initialization issue! Simulation may crash!

  Name   | Var. Value | Eqn. Mismatch
---------+------------+--------------
 v Bus 2 | 1.010      | 0.315        
 v Bus 3 | 1.010      | 0.309        

kindly let me know how can I fix this issue?
Also I want to get into more details of the values used in model so that I can set  their limits and range. for example in PVD1, on what basis values of qmx, qmn,Pmx,ialim etc are set. 
Is there any manual where in the calculation of these values (As per our system requirement) are mentioned?

**Reply by cuihantao** [reply]
Your PVD1 and ESD1 reached limit for reactive current. It’s probably because Ialim is too small. 

Unfortunately, I don’t have a manual written yet. The best way is to follow the source code.

**Comment by rana253** [comment]
Hi Dr. Cui,
Your responses have been a great help for me so far.
Need few more suggestions from your side.
I just wanted to know that is there any way to verify/ check the actual power output (in KW/MW/MVA) for the input provided in Per unit value

**Reply by cuihantao** [reply]
The power flow results can be verified in the output txt file at the bottom in the Algebraic Variable section.

**Reply by rana253** [reply]
Hi Dr. Cui,
Thank you for your response.

In 9 bus system, I need to add 3 transformers. For this, I have considered 3 transformers in "Line" (param) and made them active by putting the below values for Line 5,6,7
"trans"=1
"tap"=1
"phi"=0

After this can we consider three transformers are added or we have to do something in addition to this?

**Reply by cuihantao** [reply]
Please see the example test cases. You will at least need to specify the mandatory parameters, such as `bus1` and `bus2`, as well as parameters. You will need to study the transformer model as `tap=1` and `phi=0` makes it simply a line.

**Reply by rana253** [reply]
In Line Tab we have mentioned all the parameters including bus1, bus2. Can you please share any transformer model or any link for it so that I can refer the same.

I have gone through the ieee 14 bus excel, but there is no "transformer" tab in the spreadsheet where in all the parameters to be mentioned. Please guide to get me into right direction

**Comment by cuihantao** [comment]
You can study The PSAT manual or Power System Modelling and Scripting by Prof. Milano.

In Model references of the ANDES manual you can find the equations for the Line.

Two winding transformers share the same model as Line, so there is no separate one for transformers.

Regards,
Hantao Cui

**Comment by rana253** [comment]
Hi Dr. Cui,
Your responses have been a great help for me so far.
Just I have another doubt, how can I see slack generator generation value in simulation. let me describe little more 
 PVD1 generation = .01
 ESD1 charging= -0.005
 Load in PQ bus =0.004
remaining goes to `slack generator -0.001`, this value will be changed as per generation and load so during simulation how can I get this value?

**Reply by jinningwang** [reply]
Maybe you can plot it out for the whole curve, or just type `ss.GENROU.Pe` to see the current value.

**Reply by rana253** [reply]
Hi,
Thanks for your reply
yes I want to see current value. I have checked with `ss.GENROU.Pe` and I am getting the same value even after altering the values of PVD1 Generation and ESD1 charging while doing simulation. For example-
Initial Values:
PVD1 generation = .01
ESD1 charging= -0.005
Load in PQ bus =0.004
remaining goes to slack generator -0.001,

After altering the values while simulation:
PVD1 generation = .012
ESD1 charging= -0.006
Load in PQ bus =0.004
Now, remaining should go to slack generator i.e. -0.002, but in this case also its showing '-0.001' (initial set value)

My query is how can I get the corresponding values of Slack generator, when the values of PVD1 generation and ESD1 charging are altered?

**Reply by jinningwang** [reply]
1. Make sure those values are altered successfully, you can plot out the curve to check. Sometimes the code does not work, you can use `.set()` method.
2. If the `GENROU` is equipped with `TurbineGov`, then it should respond to the power gap. However, the governor's response may not be able to compensate the gap 100%.
3. As for `GENROU.Pe`, you can still plot the curve out to see if there is any change.
